pkg/lightning/backend/kv/session.go

```diff
@@ -47,18 +51,77 @@ func (*invalidIterator) Valid() bool {
 func (*invalidIterator) Close() {
 }
 
+type bytesBuf struct {
+	buf []byte
+	idx int
+	cap int
+}
+
+func (b *bytesBuf) add(v []byte) []byte {
+	start := b.idx
+	copy(b.buf[start:], v)
+	b.idx += len(v)
+	return b.buf[start:b.idx]
+}
+
+func newBytesBuf(size int) *bytesBuf {
+	return &bytesBuf{
+		buf: manual.New(size),
+		cap: size,
+	}
+}
+
+func (b *bytesBuf) destroy() {
+	if b != nil {
+		manual.Free(b.buf)
+		b.buf = nil
+	}
+}
+
 type kvMemBuf struct {
+	sync.Mutex
 	kv.MemBuffer
-	kvPairs []common.KvPair
-	size    int
+	buf           *bytesBuf
+	availableBufs []*bytesBuf
+	kvPairs       *KvPairs
+	capacity      int
+	size          int
+}
+
+func (mb *kvMemBuf) Recycle(buf *bytesBuf) {
+	buf.idx = 0
+	buf.cap = len(buf.buf)
+	mb.Lock()
+	mb.availableBufs = append(mb.availableBufs, buf)
+	mb.Unlock()
+}
+
+func (mb *kvMemBuf) AllocateBuf(size int) {
+	mb.Lock()
+	size = utils.MaxInt(units.MiB, int(utils.NextPowerOfTwo(int64(size)))*2)
+	if len(mb.availableBufs) > 0 && mb.availableBufs[0].cap >= size {
+		mb.buf = mb.availableBufs[0]
+		mb.availableBufs = mb.availableBufs[1:]
+	} else {
+		mb.buf = newBytesBuf(size)
+	}
+	mb.Unlock()
 }
 
 func (mb *kvMemBuf) Set(k kv.Key, v []byte) error {
-	mb.kvPairs = append(mb.kvPairs, common.KvPair{
-		Key: k.Clone(),
-		Val: append([]byte{}, v...),
+	kvPairs := mb.kvPairs
+	size := len(k) + len(v)
+	if mb.buf == nil || mb.buf.cap-mb.buf.idx < size {
+		if mb.buf != nil {
+			kvPairs.bytesBuf = mb.buf
+		}
+		mb.AllocateBuf(size)
+	}
+	kvPairs.pairs = append(kvPairs.pairs, common.KvPair{
```

Comment:
Do we need to catch `bytesBuf` to common.KvPair when `buf` is enough?

Reply:
If the current `buf` is not enough large, it seem that we discard it directly.  Would it be wasteful? I think we shall split them into different kind of buf. such as `32B, 128B, 512B,  1KB,  2KB, 4KB, 8KB,  16KB, 32KB, 64KB` and so on.

Reply:
> Do we need to catch bytesBuf to common.KvPair when buf is enough?

No. Only the last KvPairs that reference to the `buf` should take the responsibility to recycle the buffer because only this message are consumed, then the buffer is not referenced anymore and can be recycled.

> If the current buf is not enough large, it seem that we discard it directly. Would it be wasteful? I think we shall split them into different kind of buf. such as 32B, 128B, 512B, 1KB, 2KB, 4KB, 8KB, 16KB, 32KB, 64KB and so on.

yes, the current logic is to simply design. Since the default buf size is 1MiB, and 
message batch size is 64KiB, it should be large enough for most case, so in most case it should be ok. And since go is a GC language , we can depend on go GC and memory allocator to help us manage the buf allocation strategy to avoid too complicated memory management.